### PR TITLE
Release 0.4.0

### PR DESCRIPTION
**Pull request by danitso-dp**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #47
Closes #51
Closes #52
Closes #53
Closes #54
Closes #55
Closes #56
Closes #57
Closes #59
Closes #60

Release note for [CHANGELOG](https://github.com/danitso/terraform-provider-proxmox/blob/master/CHANGELOG.md):

```release-note
See references to PRs.
```

### CODE REVIEW

**Comment by danitso-dp** [comment]
@luhahn @blz-ea @patsissons @ryan4yin this is first attempt to solve merge conflicts between all your PRs. I'll continue testing this and include more of the remaining PRs. In case you test this as well and find issues, do create a PR and submit a reference to it here.

**Comment by patsissons** [comment]
oh fantastic, i should be able to give this a whirl today.

**Comment by luhahn** [comment]
Will test this on Monday :)

**Comment by ryan4yin** [comment]
greate! I'll test this today.

**Comment by luhahn** [comment]
Tested on two of our setups, looks good so far. Will check some more and review the changes here :)

**Comment by luhahn** [comment]
Oh well, "looks good so far" seems to be too early, the cloud init does not seem to work. I check into this issue now.

**Comment by ryan4yin** [comment]
> Oh well, "looks good so far" seems to be too early, the cloud init does not seem to work. I check into this issue now.

same problem here.

**Comment by luhahn** [comment]
Seems that the qcow2 image for cloud init is not created. The ide2 device for cloud init is created correctly.

**Comment by luhahn** [comment]
Found the Error, will create a PR which can be included here

**Comment by luhahn** [comment]
Okay, now it looks really good on my proxmox cluster.

Clone works with:

* multiple disks
* resize
* cloud init
* reboot
* on_boot

**Comment by danitso-dp** [comment]
@luhahn sounds good. I noticed these issues as well yesterday so I'll continue testing later today and take a final look at documentation, goreleaser etc. to ensure we can get the provider into the registry from v0.4.0 and onwards.

**Comment by ryan4yin** [comment]
I wrap this branch to use it in [pulumi](https://github.com/ryan4yin/pulumi-proxmox).
When cloning from a vm that already contains a cloudinit disk, an error occurs:

```
Updating (dev):
     Type                          Name              Status                  Info
 +   pulumi:pulumi:Stack           pve-test-dev      **creating failed**     1 error
 +   ├─ pulumi:providers:proxmox   proxmox-provider  created                 
 +   └─ proxmox:VM:VirtualMachine  test-vm-0         **creating failed**     1 error
 
Diagnostics:
  pulumi:pulumi:Stack (pve-test-dev):
    error: update failed
 
  proxmox:VM:VirtualMachine (test-vm-0):
    error: 1 error occurred:
        * creating urn:pulumi:dev::pve-test::proxmox:VM/virtualMachine:VirtualMachine::test-vm-0: Received an HTTP 500 response - Reason: lvcreate 'pve/vm-104-cloudinit' error:   Logical Volume "vm-104-cloudinit" already exists in volume group "pve"
```

File: proxmoxtf/resource_virtual_environment_vm.go [hunk 1]
```diff
@@ -3524,17 +3529,17 @@ func resourceVirtualEnvironmentVMUpdateDiskLocationAndSize(d *schema.ResourceDat
...
+					Disk:               diskOldInterface,
 					TargetStorage:      diskNewDatastoreID,
 				})
 			}
 
 			diskOldSize := diskOldBlock[mkResourceVirtualEnvironmentVMDiskSize].(int)
 			diskNewSize := diskNewBlock[mkResourceVirtualEnvironmentVMDiskSize].(int)
 
-			if diskOldSize != diskNewSize {
+			if diskOldSize <= diskNewSize {
```
Comment (ryan4yin) [inline proxmoxtf/resource_virtual_environment_vm.go]: Should this be changed to `<`?

File: proxmox/virtual_environment_vm.go [hunk 2]
```diff
@@ -130,29 +157,194 @@ func (c *VirtualEnvironmentClient) GetVMStatus(nodeName string, vmID int) (*Virt
...
+	if strings.Contains(err.Error(), "you can't move to the same storage with same format") {
+		// if someone tries to move to the same storage, the move is considered to be successful
+		return nil
+	}
+
+	if err != nil {
+		return err
+	}
+
+	err = c.WaitForNodeTask(nodeName, *taskID, 86400, 5)
```
[the remark this answers is not in the document]
Comment (luhahn) [reply proxmox/virtual_environment_vm.go:173]: mh yes, maybe thats a good idea. I will prepare a patch tomorrow

**Comment by danitso-dp** [comment]
I'll get back to testing and preparing this release on Sunday/Monday. I'll keep an eye out for PRs in the meantime so they can be merged once they're made available.

File: proxmox/virtual_environment_vm.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ryan4yin) [reply proxmox/virtual_environment_vm.go:173]: @luhahn thanks, I'll test it tomorrow.

**Comment by ryan4yin** [comment]
`parameterize-timeouts` works fine.

**Comment by luhahn** [comment]
> 
> 
> `parameterize-timeouts` works fine.
> but the error I mentioned above still exists: `Logical Volume "vm-104-cloudinit" already exists in volume group "pve"`, and I have no idea about how to fix it.

@ryan4yin  What would the expected behaviour here? You said, that there already was a cloudinit disk. You want multiple cloudinit files?

**Comment by ryan4yin** [comment]
@luhahn when clone from a template vm that already contains a cloudinit disk, I hope the proxmox provider can just skip the creation of cloudinit disk.

**Comment by luhahn** [comment]
I'll check if i can patch that :)

**Comment by ryan4yin** [comment]
@luhahn thanks :+1:

**Comment by ryan4yin** [comment]
> I'll check if i can patch that :)
> 
> But you didn't define the initialization {} field, right?

no, I  defined the initialization. the configuration I used: https://github.com/ryan4yin/pulumi-proxmox#examples

**Comment by luhahn** [comment]
Mh, if initialization is defined the provider will try to create the cloudinit file. Without it won't.
I check what can be done.

**Comment by luhahn** [comment]
@ryan4yin  Ok, I had a look at the code. And the problem with existing cloud init files is, that we currently have cloud-init and cdrom on fixed IDE devices. Cloud init on ide2 and cd on ide3. However if your vm which you are cloning from already has a cloud init device and you want the provider to dynamically use that device instead of the fixed ones. We need to change alot of the current logic.

This is because terraform providers are using read and update function in order to determine if changes have been made, so we need to make quite a few changes in the code.

Since I already had a look I tried to set the interfaces for cd and cloudinit dynamically, like the disk interfaces. I also added some checks to see if a cloud init device is already existing. But i am stuck in how to update those correctly without terraform deleting them all the time.

Another idea would be to delete the current cloud init device, which has been received from your clone. And override the settings with the ones from your initialization configuration.

The last solution would be that you simply remove the initialization config..

**Comment by ryan4yin** [comment]
@luhahn thanks, I got it. It's not a big problem, now I solve this by manually deleteing the cloudinit disk from the template VM.

**Comment by luhahn** [comment]
The current state of this PR is running in our test/development cluster for quite some time now. Except for the existing cloud-init i couldn't notice any bugs for cloning :)

**Comment by mleone87** [comment]
Hi, first comment on this provider!

I downloaded the PR code and build it to my test setup. I use the following resource definition, partially modified from the upsteream documentation(mainly removing obsolete tf syntax).
The problem is that the resource is constantly deleted and recreated, with or without any modification

```
resource "proxmox_virtual_environment_vm" "ubuntu_vm" {
  name        = "vm1"
  description = "Managed by Terraform333"

  node_name = "node01"
  vm_id     = 4321

  agent {
    enabled = true
  }

  cpu {
    type = "host"
    cores = 2
  }

  disk {
    interface    = "virtio"
    datastore_id = "nvme250"
    file_id      = proxmox_virtual_environment_file.ubuntu_cloud_image.id
  }

  initialization {
    datastore_id = "nvme250"
    ip_config {
      ipv4 {
        address = "dhcp"
      }
    }

    user_account {
      keys     = [trimspace(tls_private_key.ubuntu_vm_key.public_key_openssh)]
      password = "ubuntu"
      username = "ubuntu"
    }
  
  }

  network_device {}

  operating_system {
    type = "l26"
  }

  serial_device {}
}

resource "proxmox_virtual_environment_file" "ubuntu_cloud_image" {
  content_type = "iso"
  datastore_id = "local"
  node_name    = "node01"

  source_file {
    path = "~/ubuntu1804.img"
  }
}

resource "tls_private_key" "ubuntu_vm_key" {
  algorithm = "RSA"
  rsa_bits  = 2048
}


```

**Comment by luhahn** [comment]
can you provide more information, like logs?

**Comment by mleone87** [comment]
> can you provide more information, like logs?

sure, if you want I can upload the full copy-paste of my terminal somewhere the full logs but this is the terraform plan that is subsequent to a successful apply.

```
terraform plan
proxmox_virtual_environment_file.ubuntu_cloud_image: Refreshing state... [id=local:iso/ubuntu1804.img]
tls_private_key.ubuntu_vm_key: Refreshing state... [id=bb9473be8766c8e0d60f1097b385a98a446efc80]
proxmox_virtual_environment_vm.ubuntu_vm: Refreshing state... [id=4321]

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
-/+ destroy and then create replacement

Terraform will perform the following actions:

  # proxmox_virtual_environment_vm.ubuntu_vm must be replaced
-/+ resource "proxmox_virtual_environment_vm" "ubuntu_vm" {
      ~ id                      = "4321" -> (known after apply)
      ~ ipv4_addresses          = [
          - [
              - "127.0.0.1",
            ],
          - [
              - "192.168.10.98",
            ],
        ] -> (known after apply)
      ~ ipv6_addresses          = [
          - [
              - "::1",
            ],
          - [
              - "fe80::4497:e4ff:fe7d:2024",
            ],
        ] -> (known after apply)
      ~ mac_addresses           = [
          - "00:00:00:00:00:00",
          - "46:97:E4:7D:20:24",
        ] -> (known after apply)
        name                    = "vm1"
      ~ network_interface_names = [
          - "lo",
          - "eth0",
        ] -> (known after apply)
        # (17 unchanged attributes hidden)


      ~ cpu {
          - flags        = [] -> null
            # (6 unchanged attributes hidden)
        }

      ~ disk {
          + file_id      = "local:iso/ubuntu1804.img" # forces replacement
          ~ interface    = "scsi0" -> "virtio"
            # (3 unchanged attributes hidden)
        }

      ~ initialization {
            # (1 unchanged attribute hidden)

          ~ ip_config {
              ~ ipv4 {
                    # (1 unchanged attribute hidden)
                }
            }

          ~ user_account {
              ~ password = (sensitive value)
                # (2 unchanged attributes hidden)
            }
        }

      ~ network_device {
          - mac_address = "46:97:E4:7D:20:24" -> null
            # (5 unchanged attributes hidden)
        }


        # (3 unchanged blocks hidden)
    }

Plan: 1 to add, 0 to change, 1 to destroy.

Changes to Outputs:
  ~ ubuntu_vm_ip = [
      - [
          - "127.0.0.1",
        ],
      - [
          - "192.168.10.98",
        ],
    ] -> (known after apply)

------------------------------------------------------------------------
```


I can understand why terraform would destroy a vm created that way since iit seems correct from it's perspective, but then I can not figure out how we could create a new vm(not a clone) with an image uploaded that way.

**Comment by danitso-dp** [comment]
I've finally managed to finish a pile of very important tasks, so I'll be dedicating a lot of time next week to get v0.4.0 released.

Based on your testing and what I managed to test earlier, it seems we're close to a stable version so I won't pull in more PRs. Instead, I'll begin work on the v0.5.0 release next week once v0.4.0 is released.

**Comment by danitso-dp** [comment]
I can't find any major issues with this release so I'll take one last look at the documentation and changelog before releasing it. I'll also push the final changes to the release flow before merging this PR.

**Comment by danitso-dp** [comment]
Merging this now. Any issues will be fixed against the master branch.

**Pull request merged**